### ISSUE TITLE
Library implications on memory and performance + high-performance computing capabilities

### ISSUE THREAD

**Issue by pcanas** [chaquo/chaquopy#653]
Hi Malcolm,

I was wondering if there is any documentation or reference values on the implications of using Chaquopy in terms of:
* Performance: 
   * Does the library impact the speed execution on other parts of the app? 
   * How long does it take to start the Python instance and the module?
   * How longer/shorter does it take to load depending on the dependencies included in the build script?
   * How fast is it to call a python script versus implementing the same functionality with native code?

* Memory:
   * How much space does the chaquopy library (alone) take?
   * How is exactly the storage space allocated in terms of 'Size of the app' and 'User data'? In other words, if I keep adding dependencies, is it going to impact one or the other?

* I was also wondering what the high-performance computing capabilities of the library are. More specifically, let's say that I want to use numpy which internally makes multithreading. Am I going to obtain the same multithreading capabilities when running the same script via Chaquopy?

Let me know if any question was unclear. And thanks for your help!

**Comment by mhsmith** [comment]
>   * Does the library impact the speed execution on other parts of the app?

No: the library only runs when you call it, it doesn't do anything in the background.

>   * How long does it take to start the Python instance and the module?

Starting Python takes about 2 seconds on a mid-range device. The time taken to load a module will of course depend on what the module does. 

I suggest you take one of the [examples](https://chaquo.com/chaquopy/doc/current/examples.html), modify it to simulate your use case, and see what timings you get. If you feel it's too slow, please comment here with details of what your code does and what device you're using.

>   * How longer/shorter does it take to load depending on the dependencies included in the build script?

Chaquopy unpacks as little as possible from the APK, so adding Python code and libraries usually doesn't make a difference to how quickly Python itself starts. The only exception I can think of is if you have some large data files in your Python source directory. As it says in [the documentation](https://chaquo.com/chaquopy/doc/current/android.html#android-data):

> Data files within a top-level package (i.e. a top-level directory containing an `__init__.py` file) will not be available until the first time that package is imported. All other data files will be available as soon as Python has started.

So for example, if you have a large data file at the root of your source tree, it will be extracted from the APK into the device filesystem as soon as Python is started, which may cause a delay. However, this will only happen the first time your app is started. Subsequent runs will reuse the same file until the app is updated.

>   * How fast is it to call a python script versus implementing the same functionality with native code?

Pure Python code is usually slower than Java/Kotlin. So if you're processing large amounts of data then you should use libraries like NumPy and SciPy, which come with optimized algorithms in C/C++/Fortran.

I'll get back to you later on the other questions.

**Comment by pcanas** [comment]
All is clear for the moment. Thanks a looking forward to the rest!

**Comment by mhsmith** [comment]
> How much space does the chaquopy library (alone) take?
> How is exactly the storage space allocated in terms of 'Size of the app' and 'User data'? In other words, if I keep adding dependencies, is it going to impact one or the other?

Chaquopy increases APK size by about 3.5 MB, plus 5 MB per <a href="https://chaquo.com/chaquopy/doc/current/android.html#android-abis">ABI</a>. When installed on a device, it requires about 12.5 MB of user data storage on top of the APK size.

Adding Python libraries mostly affects the APK size, but any data files or native libraries (.so files) will have to be extracted onto the device, so they'll increase the user data size as well.

As it says in [the FAQ](https://chaquo.com/chaquopy/doc/current/faq.html#faq-size), the easiest way to reduce all of these is to reduce the number of ABIs in your app.

> I was also wondering what the high-performance computing capabilities of the library are. More specifically, let's say that I want to use numpy which internally makes multithreading. Am I going to obtain the same multithreading capabilities when running the same script via Chaquopy?

In general, yes: our library builds should be able to take advantage of whatever multi-core hardware your device has. However, we don't currently support any GPU acceleration in TensorFlow or PyTorch.

**Comment by pcanas** [comment]
Thanks, Malcolm!
